[NativeAOT] Fix Lock's usage of NativeRuntimeEventSource.Log to account for recursive accesses during its own class construction

kouvel opened the pull request:
`NativeRuntimeEventSource.Log` can be null if it's being constructed in the same thread earlier in the stack. Added null checks.

Fixes https://github.com/dotnet/runtime/issues/94728

Commit:
Fix Lock's usage of NativeRuntimeEventSource.Log to account for recursive accesses during its own class construction

- `NativeRuntimeEventSource.Log` can be null if it's being constructed in the same thread earlier in the stack, added null checks

Fixes https://github.com/dotnet/runtime/issues/94728

MichalStrehovsky commented:
With this and #94241, what is our confidence that `Lock` can still safely be used within the implementation of the native AOT runtime? Do we need to consider migrating these low level usages to `LowLevelLock`?

In #94515 I used `LowLevelLock` not because I hit some issue, but I just couldn't reason if it's safe to use `Lock` in the implementation of the C# `typeof` (or `object.GetType`).

VSadov commented:
> With this and https://github.com/dotnet/runtime/pull/94241, what is our confidence that Lock can still safely be used within the implementation of the native AOT runtime? Do we need to consider migrating these low level usages to LowLevelLock?

When we were stabilizing the old Lock the same question was asked. I think the same reasoning still applies - 
NativeAOT runtime is written in C#, uses off-the shelf components and calls everchanging set of various things. It would be impractical to require that nothing of that uses locks or must use a special kind of lock. On the other hand, the Lock is a very simple mechanism, with a very simple purpose and calls a very limited set of APIs. It is definitely possible to ensure that those APIs do not use Lock and thus guarantee that there is no reentrancy. 

Old lock did not have this problem, ... eventually. The implementation in #93879 does not seem to be affected. This is a fixable issue.

> In https://github.com/dotnet/runtime/pull/94515 I used LowLevelLock not because I hit some issue, but I just couldn't reason if it's safe to use Lock in the implementation of the C# typeof (or object.GetType).

The only legit use of LowLevelLock in NativeAOT is in implementation of Wait subsystem. Since Lock uses that, we can't use Lock there.
In most other places, including `typeof`, it should be fine to use Lock.

VSadov commented:
@kouvel - is there a good understanding of the repro? It worries me a bit that the bug reproduces exclusively on linux-arm64. That was also the case with the previous bug.
Is linux-arm64 the only configuration that enables logging, by any chance?  

Relying on linux-arm64 for stress might not be a good place to be. If there was a way to force this to repro on other platforms, especially on win-x64, it would be very valuable. I wonder what makes linux-arm64 special here.

MichalStrehovsky commented:
> The only legit use of LowLevelLock in NativeAOT is in implementation of Wait subsystem. Since Lock uses that, we can't use Lock there.
> In most other places, including typeof, it should be fine to use Lock.

Thanks, I sent out a PR to use that instead: #94991.

> Is linux-arm64 the only configuration that enables logging, by any chance?

It shouldn't be linux-arm64 specific, but we definitely enable EventSource for fewer tests. EventSource is disabled by default in the shipping configuration (some workloads like ASP.NET will enable it back). The tests mirror that. Only tests that test event source enable it back (search for `<EventSourceSupport>true</EventSourceSupport>` to see the tests that do).

kouvel commented:
> @kouvel - is there a good understanding of the repro? It worries me a bit that the bug reproduces exclusively on linux-arm64. That was also the case with the previous bug.

It's a timing issue that can lead to the stack trace in https://github.com/dotnet/runtime/issues/94728 or something like it. It's not specific to a platform or architecture, probably not easy to repro without inducing timings.

> With this and https://github.com/dotnet/runtime/pull/94241, what is our confidence that `Lock` can still safely be used within the implementation of the native AOT runtime?

There appears to be a circular dependency where Lock does some lazy initialization that involves class construction, which in turn uses Lock for synchronization. That makes the system a bit brittle to changes. For instance, new issues were uncovered when raising events through `NativeRuntimeEventSource`. Maybe one way to break that circular dependency would be for runtime uses of Lock to pass in a boolean that bypasses lazy initialization and uses defaults, along with not raising events. I'll see if we see more of this type of issue or cases that are not easily fixable from the Lock side, maybe something like that can be done.

VSadov commented:
`RuntimeEventSource` is eagerly initialized on module load because it has `[ModuleInitializer]`. That indirectly initializes `NativeRuntimeEventSource`. By the time we start running some meaningful code, `NativeRuntimeEventSource` should be initialized.  

Somehow on arm64 we start seeing contentions from sockets before module initialization is complete. Not sure if that was expected by whoever have put `[ModuleInitializer]` there.

VSadov commented:
I think the most reliable fix for this would be to stop calling static initialization from the Lock entirely, and make something else do the initialization instead.  
For that we need: 
1. make static initialization optional for the Lock functionality. It may result in losing a few events when timing is tight, but it does not seem to be a big deal or avoidable. 
(currently suggested change is half way there already - it allows losing events while the Log is not ready)  
1. make something else to perform initialization of statics. There are a few options here:
-- Use [ModuleInitializer] 
-- Use finalizer of some dummy object 
like in https://github.com/dotnet/runtime/blob/1c4bd555ecca1ed49cd7a259f23c35e6d83665ee/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Lock.NativeAot.cs#L130-L146

In either case there is a chance to have a small window of operating with uninitialized statics, but that is ok as long as the initialization is optional.  

The advantage of [ModuleInitializer] is being more explicit and upfront, but there could be concern that module init would run at program startup and the NativeRuntimeEventSource initialization could be a bit heavy (there appears to be some reflection use, type loading, attribute parsing,..). 
I am not sure about the cost of reading/parsing the config settings.

The Finalizer approach allows to wait until we are sure that we have a program that may have contentions to report.

MichalStrehovsky commented:
> Somehow on arm64 we start seeing contentions from sockets before module initialization is complete. Not sure if that was expected by whoever have put [ModuleInitializer] there.

Maybe an ordering issue with this: https://github.com/dotnet/runtime/blob/784537a9cdc97bc5f45f6606f3f9837a7f755236/src/libraries/Common/tests/System/Net/QuicLoad.cs#L12-L20

The code looks to be Linux-only so that would explain we don't see it on Windows. Cc @LakshanF. IIRC RuntimeEventSource initialization it was made a ModuleInitializer because running it as part of CoreLib initialization was too early and we don't have any other initialization phase before Main. Maybe we need one.

jkotas commented:
We do topological sort of ModuleInitializers. CoreLib module initializer should run first and the Quic module initializer should run second. Both CoreLib module initializer and the Quic module initializer start new threads. The problem seems to be triggered when the background work done by the Quic module initializer happens faster than the background work done by the Quic module initializer.

I think the root cause of the problem is use of managed `Lock` by the native AOT ClassConstructorRunner. The equivalent code in regular CoreCLR uses low-level lock and it is why it avoids the dependency cycles. Should we fix this issue by switching ClassConstructorRunner to use the low-level lock?

kouvel commented:
> Second related problem is that the managed locks use high-level wait that can be overridden using synchronization context.

Given that it wouldn't be feasible to use `Lock` in low-level cases, `Lock` could be simplified a bit by removing the lazy initialization and maybe other things.

Should I make the relevant changes to ClassConstructorRunner also?

jkotas commented:
> Should I make the relevant changes to ClassConstructorRunner 

I think so. And delete/simplify code that has been working around the Lock use in ClassConstructorRunner. Thank you!

VSadov commented:
> We do topological sort of ModuleInitializers. CoreLib module initializer should run first and the Quic module initializer should run second. Both CoreLib module initializer and the Quic module initializer start new threads. The problem seems to be triggered when the background work done by the Quic module initializer happens faster than the background work done by the Quic module initializer.

Is that a problem or something expected?
If they both start threads and complete in random order, why bother ordering them? Is it the order of their start or completion that matters?  
(I do not have a lot of opinion on this, but it kind of still feel suspicious that it affects only linux-arm64)

jkotas commented:
> >  The problem seems to be triggered when the background work done by the Quic module initializer happens faster than the background work done by the Quic module initializer.

> Is that a problem or something expected?

It is expected. 

The problem is that two threads running in parallel cause contention in constructor runner that leads to the crash.

VSadov commented:
> Second related problem is that the managed locks use high-level wait that can be overridden using synchronization context. Try running this:

It is a bit surprising why we decided to do this (perhaps something with COM/STA as usual), and keep maintaining. In normal use Lock has 2 levels of optimizations to avoid Waiting, and `lock{}` has 3 levels. Outside of deadlocks the feature use would be very nondeterministic. 
It might be late to argue that this feature is as useful as `Thread.Interrupt` and would not be very missed in the same way.

jkotas commented:
> How far will this go?

Yes, it needs to stop somewhere. The split between ordinary locks and managed locks is very arbitrary today. My point is that the best way to fix reentrancy issues specific to native AOT is by using same type of lock between runtimes in the current state of the world.

> I think it is doable and is a lot cheaper to just keep lock{}/Lock compatible with the runtime except for a set of functionality that Lock itself uses (and that set can be very small).

I would be open to the idea of making the new Lock use primitive waits. It would require investigation of what SynchronizationContext overrides are used for today and whether they would get broken by it.

VSadov commented:
> I would be open to the idea of making the new Lock use primitive waits. It would require investigation of what SynchronizationContext overrides are used for today and whether they would get broken by it.

Right. I was going to literally suggest "I'd go as far as use uninterruptible Wait in Lock", but decided to search for anything that could be possibly broken. (other than tests).

So far I found nothing interesting. SynchronizationContext is used, but mostly for Post/Send/Complete purposes. 

Noone seems to be overriding Wait in any nontrivial way. I found only 2 cases both just forward to nonalertable WaitForMultipleObjectsEx, ironically - to suppress the message pumping behavior of standard wait.

VSadov commented:
> Yes, it needs to stop somewhere. The split between ordinary locks and managed locks is very arbitrary today. My point is that the best way to fix reentrancy issues specific to native AOT is by using same type of lock between runtimes in the current state of the world.

I think the root cause for reentrancy is that Lock initializes EventSource as a part of its regular operations and that runs arbitrary code and that causes stress regressions. If that initialization happens on a side - in module initializer or in a finalizer, the problem will disappear. 

I do not think that CoreCLR is a good guidance on what lock to use. The reasons in CoreCLR could be different, often just because of different native/managed split or some appdomain/sql stuff that may not even be applicable anymore. 
In NativeAOT most locks can be managed. I did not see a convincing reason other than the issue above, and that has a simpler solution than making specialized locks.

The static initialization is actually a good example. 

LowLevelLock lacks functionality for that. It also has a fairly poor performance. That was never an issue as LowLevelLock was basically a "backstop" lock for cases that happen rarely or nearly always resulted in blocking wait. Perf improvements would not "move the needle".  
For static initializers we would need a recursive lock that also supports timeouts and preferably has the best performance that we can have. We may end up with basically a copy of Lock, just one that uses nonblocking Wait and does not do EventSource diagnostics. 

It would be a roundabout solution to something that could be solved inside the Lock itself.

Alternatively, Lock could have an internal ctor/flag to not initialize EventSource, but I think plugging into the diagnostics would be nice even for system locks.

jkotas commented:
> For static initializers we would need a recursive lock that also supports timeouts

I do not think we need that for static initializers. The static initializers are scheduled by "dead-lock aware lock". The regular (low-level) lock is needed to protect the dead-lock aware lock state. There should not be any recursion. The low-level lock for this purpose does not need to be fine-tuned, given other overheads involved in running the static constructors. 

We have the low-level lock exposed by the System.Native PAL. I think it would be hard to get rid of all uses of it. As long as we have the low-level lock exposed by the System.Native PAL, it should be fine to use it in S.P.CoreLib as needed.

VSadov commented:
> There should not be any recursion. The low-level lock for this purpose does not need to be fine-tuned, given other overheads involved in running the static constructors.

Given the fact that we run into contentions, and often enough to bring up the reentrancy issue, indicates that this lock can be a hot spot and thus a potential bottleneck if lock does not handle contentions well.

> As long as we have the low-level lock exposed by the System.Native PAL, it should be fine to use it in S.P.CoreLib as needed.

It is ok to use it. It works. It is just generally there are no reasons to use it vs. regular `lock{}`/`Lock`. 
Perhaps people get some sense of extra safety - "oh oh, this looks like low level code, I probably should use LowLevelLock".

There is an issue of "what if someone does things with SynchronizationContext.Wait", but to shield from that most locks would need to be LowLevelLock. That is probably why noone "does things with SynchronizationContext.Wait" as that would be very breaking.

kouvel commented:
What about locks taken on UI threads? I imagine they would still need to use a message-pumping wait for correctness and that could run user code and potentially lead to reentrancy.

kouvel commented:
Given that class construction runs user code, I'm figuring that the lock used for synchronizing it should have the same behavior. As for SynchronizationContext wait overrides and Thread.Interrupt, they are things that should probably be recommended against, and maybe even deprecated eventually. I'm not sure though what kind of breaks would be involved from these low-level kind of uses of interruptible/overridable/message-pumping waits.

kouvel commented:
Not sure though if a lock is actually held around running user code during class construction. If not, then it could very well be a simpler lock.

jkotas commented:
> Not sure though if a lock is actually held around running user code during class construction.

There is a lock held around running user static constructor code, but the attempt to run static constructor recursively needs to be short-circuited: https://github.com/dotnet/runtime/blob/57e3b0b75b9799f4b62e3a73fa81cc7c7fd671e0/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/ClassConstructorRunner.cs#L145-L146 . It means that you need to be able to check whether the lock is taken by current thread, but it is easy to on the side using thread id if the lock does not support it.

I do not know why the native AOT ClassConstructorRunner uses timeouts. It looks both inefficient and complicated.

VSadov commented:
> What about locks taken on UI threads? I imagine they would still need to use a message-pumping wait for correctness and that could run user code and potentially lead to reentrancy. Should locks used in the runtime (in low-level situations) also use message-pumping waits?

With runtime is implemented mostly in managed code it a simpler model to see it as just extension of user's code - just another library. There are some exceptions for things too close to GC, but mostly it is code that the app could contain.

For the message pumping - I'd guess by default the same considerations equally apply to user code vs. managed runtime and sorting locks onto which may not need to pump might not make a lot of difference overall.

> As for SynchronizationContext wait overrides and Thread.Interrupt, they are things that should probably be recommended against, and maybe even deprecated eventually. 

I agree with this. 

Thread.Interrupt is actually not supported on NativeAOT, but on the long term TODO list. Might be a good time to just deprecate it. I wonder why it was not deprecated at the time Thread.Abort was.

jkotas commented:
> I wonder why it was not deprecated at the time Thread.Abort was.

It is popular in code ported from Java. According to https://apisof.net/catalog/4edbcb9f-8e93-65b1-23e2-49877d968c90, it is used by 12.4% of apps in our appcompat test suite.

jkotas commented:
> With runtime implemented mostly in managed code it a simpler model to see it as extension of user's code - just another library. There are some exceptions for things too close to GC, but mostly it is code that the app could contain.
> For the message pumping - I'd guess by default the same considerations equally apply to user code vs. managed runtime and sorting locks onto which may not need to pump might not make a lot of difference overall.

This is correct until things go wrong. Things go wrong with the recursive message pumping all the time. I think it is important that the lowest levels of the runtime are immune to this mess so that it manifest as an exception that the app developer can understand.

rickbrew commented:
> > As for SynchronizationContext wait overrides and Thread.Interrupt, they are things that should probably be recommended against, and maybe even deprecated eventually.
> 
> I agree with this.

I use a custom `SynchronizationContext` for the reason listed earlier -- to stop pumping during certain regions of code that are sensitive to reentrancy. Deprecating/removing that without a suitable replacement would be an adoption blocker for me. I don't use NativeAOT, however.

hamarb123 commented:
> > As for SynchronizationContext wait overrides and Thread.Interrupt, they are things that should probably be recommended against, and maybe even deprecated eventually.
> 
> I agree with this.

I use `Thread.Interrupt`, removing it would break me. If the current `Thread.Interrupt` solution removed, it would be good to add a replacement for it, where you could `WaitInterrupt` explicitly or similar - this removes the issue of exceptions being thrown in random places, and would be nicer and more reliable to use overall.

> Thread.Interrupt is actually not supported on NativeAOT

If we remove the current `Thread.Interrupt` solution, hopefully we could design one that works on .NET 9 (with NativeAOT also).

kouvel commented:
> If the current `Thread.Interrupt` solution removed, it would be good to add a replacement for it, where you could `WaitInterrupt` explicitly or similar - this removes the issue of exceptions being thrown in random places, and would be nicer and more reliable to use overall.

Agreed, I feel `Thread.Interrupt` can be a bit disruptive, perhaps a better option would be to offer an interruptible wait explicitly so that it can be more clearly controlled which wait is intended to be interrupted.

> I use a custom `SynchronizationContext` for the reason listed earlier -- to stop pumping during certain regions of code that are sensitive to reentrancy. Deprecating/removing that without a suitable replacement would be an adoption blocker for me. I don't use NativeAOT, however.

Yep that's reasonable, maybe the wait overrides would be difficult to avoid, the actual wait may also be in a different library that is not in the user's control. The runtime may need to be careful about using such overridable waits in some cases.

jkotas commented:
> I feel Thread.Interrupt can be a bit disruptive,

Thread.Interrupt can be used safely only if you are certain that the code running on the other thread expects it.

> Thread.Interrupt is actually not supported on NativeAOT

If we find that Thread.Interrupt is needed in NativeAOT, I do not see any fundamental reason why it cannot be implemented in the current shape. It is "just work".

jkotas commented:
@kouvel Do you have a plan in mind for how to proceed on this one?

kouvel commented:
> @kouvel Do you have a plan in mind for how to proceed on this one?

Yes. I'm more in favor of reusing `Lock` for this purpose with a wait that does not allow reentrance. `LowLevelLock` (or some variant of it) would certainly be an option, but it's wasn't meant for this kind of purpose in its original or current state (which are the same). There might also be more uses for non-reentrant locks in other cases. `LowLevelLock` was created for specific purposes that are likely still valid. I'd like to see `LowLevelLock` go away eventually though, ideally `Lock` should be able to handle those cases just as well.

MichalStrehovsky commented:
@kouvel do you have any timelines for this? It is a heavy hitter in the CI (I see this failing every day) and we're about to snap .NET 9 Preview 1.

kouvel commented:
> @kouvel do you have any timelines for this? It is a heavy hitter in the CI (I personally see this failing every day and CI stats say we hit it 15 times a day on average) and we're about to snap .NET 9 Preview 1.

The proposed change in this PR would I believe fix the CI issue. The discussion in this PR has evolved into resolving a larger issue, which I'm currently looking into, hope to have something out soon. After thinking about it more, I may continue to employ the same technique for the CI issue, as using a finalizer doesn't guarantee timely initialization and I'd prefer to avoid going down the wait path before initialization is complete to avoid other potential reentrancy issues (and to have a way to fix a reentrancy issue on that path if one is introduced), though I'm still evaluating it.

jkotas reviewed (approved):
Thanks

jkotas commented:
> resolving a larger issue, which I'm currently looking into, hope to have something out soon. 

Could you please open an issue for this if there is not one already?

Commit:
Add a null check for another access to NativeRuntimeEventSource.Log in the wait path

kouvel commented:
Rebased, and added a commit with one more null check in the wait path used by Lock. Opened https://github.com/dotnet/runtime/issues/97105 to track making some of the waits unalertable.

kouvel commented:
Looks like all remaining CI failures are known issues.
